### PR TITLE
#96 - Extracted Bucket from MultipartUpload

### PR DESCRIPTION
**Pull request by olegmoz**
Closes issue #96
Solves complexity in MultipartUpload by extracting Bucket class as new layer of abstraction over S3Client

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 I'm idle, could you please assign me?

**Comment by 0crat** [comment]
This pull request [#109](https://github.com/artipie/asto/issues/109) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/2aae532b-e401-4196-a74b-eeb93e037393); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by olegmoz** [comment]
@paulodamaso all `Bucket` methods are tested indirectly in `S3StorageTest` I belive that is enough coverage

**Comment by paulodamaso** [comment]
@olegmoz Great, sho please split these tests from `S3StorageTests` and put them into `BucketTests`

**Comment by olegmoz** [comment]
@paulodamaso that would be a bad idea. `Bucket` is part of `S3Storage` implementation, I'd prefer to keep test for public method, not for it's implementation details

**Comment by paulodamaso** [comment]
@olegmoz If we have a new abstraction (`Bucket`), then we shoukld test it isolated from the rest of the application; if we can't do it, means that the classes are too coupled. @g4s8 WDYT of my request for creating `BucketTests`?

**Comment by paulodamaso** [comment]
@g4s8 Please take a look at https://github.com/artipie/asto/pull/109#issuecomment-599608783

**Comment by g4s8** [comment]
@paulodamaso @olegmoz I think more tests is better than less. But we don't have coverage threshold yet.  So I'd add coverage rates limits first to enforce these metrics.

**Comment by olegmoz** [comment]
@g4s8 @paulodamaso added tests for `Bucket` class

File: src/test/java/com/artipie/asto/s3/BucketTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                    .key(key)
+                    .uploadId(id)
+                    .build()
+            )
+        ).join();
+        final byte[] downloaded;
+        try (S3Object s3Object = client.getObject(this.name, key)) {
+            downloaded = ByteStreams.toByteArray(s3Object.getObjectContent());
+        }
+        MatcherAssert.assertThat(downloaded, Matchers.equalTo(data));
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/s3/BucketTest.java]: @olegmoz  Please use `new isEqual` here

File: src/test/java/com/artipie/asto/s3/BucketTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        this.bucket().abortMultipartUpload(
+            AbortMultipartUploadRequest.builder()
+                .key(key)
+                .uploadId(id)
+                .build()
+        ).join();
+        final List<MultipartUpload> uploads = client.listMultipartUploads(
+            new ListMultipartUploadsRequest(this.name)
+        ).getMultipartUploads();
+        MatcherAssert.assertThat(uploads, new IsEmptyIterable<>());
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/s3/BucketTest.java]: @olegmoz You can inline `uploads` here, avoiding the variable creation

File: src/test/java/com/artipie/asto/s3/BucketTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    private Bucket bucket() {
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/s3/BucketTest.java]: @olegmoz Wouldn't be better to create the bucket instance in `setUp` method?

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) this job was assigned to you 6days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by olegmoz** [comment]
@paulodamaso thank you, all good points. Made the changes, please review

**Comment by paulodamaso** [comment]
@g4s8 Please merge

**Pull request merged**